keps/sig-scheduling/20190221-even-pods-spreading.md

```diff
@@ -0,0 +1,423 @@
+
+- if CA is enabled, a new machine will be provisioned to hold the new pod
+  (although old replicas will be deleted afterwards) (see
+  [#40358](https://github.com/kubernetes/kubernetes/issues/40358))
+- if CA is not enabled, it's a deadlock since the replacement pod can't be
+  placed. The only workaround at this moment is to update app strategyType from
+  "RollingUpdate" to "Recreate".
+
+Neither of them is an ideal solution. A promising solution is to give user an
+option to trigger "toleration" mode when the cluster is out of resource. Then in
+aforementioned example, a third pod is "tolerated" to be put onto node1 (or
+node2). But keep it in mind, this behavior is only triggered upon resource
+shortage. For a 3-nodes cluster, the third pod will still be placed onto node3
+(if node3 is capable).
+
+### Risks and Mitigations
+
+The feature requires additional processing for pods that use it and it is ok to
+have some performance overhead. But we will make sure our implementation will
+not have any performance penalty for pods that do not use this feature.
+
+## Design Details
+
+### API
+
+A new structure called `SpreadingConstraint` is introduced which acts as a
+standalone spec and is applied to `pod.spec`. It's only effective when it's not
+nil.
+
+```go
+type PodSpec struct {
+    // SpreadingConstraint describes how a group of pods are spread
+    // If specified, scheduler will enforce the constraint
+    // +optional
+    SpreadingConstraint *SpreadingConstraint
+    ......
+}
+```
+
+#### Option 1
+
+Inside `SpreadingConstraint`, we need hard affinityTerms (similar with
+`PodAffinityTerm`) and soft affinityTerms (similar with
+`WeightedPodAffinityTerm`). This describes when we perform even distribution,
+which pods are considered as a group.
+
+```go
+type SpreadingConstraint struct {
+    // MaxSkew describes the degree of imbalance of pods spreading.
+    // It's the max difference between the number of matching pods in any two
+    // topology domains of a given topology type.
+    // Default value is 1 and 0 is not allowed.
+    MaxSkew int32
+    // TopologyKey defines where pods are placed evenly
+    TopologyKey string
+    // Similar with the same field in PodAffinity/PodAntiAffinity
+    // +optional
+    RequiredDuringSchedulingIgnoredDuringExecution []PodAffinityTerm
+    // Similar with the same field in PodAffinity/PodAntiAffinity
+    // +optional
+    PreferredDuringSchedulingIgnoredDuringExecution []WeightedPodAffinityTerm
+}
+```
+
+#### Option 2
+
+Another option is to flatten "required" and "perferred" podAffinityTerms, and
+eliminate embedded "TopologyKey":
+
+```go
+type ScheduleDecision string
+
+const (
+    // do not schedule a pod in all circumstances
+    DoNotSchedule ScheduleDecision = "DoNotSchedule"
+    // schedule a pod despite of any circumstance
+    ScheduleAnyway ScheduleDecision = "ScheduleAnyway"
+)
+
+type SpreadingConstraint struct {
```

Comment:
Naming wise we generally don't use `ing` to describe things (it's harder to understand in English than shorter forms).

You might want to consider:

```
TopologySpreadConstraint
SpreadConstraint
PlacementSpreadConstraint
TopologyVarianceConstraint
SpreadToleration
TolerationVariance
TopologyTolerationConstrant
```

Reply:
Thanks @smarterclayton. I think `TopologySpreadConstraint` is very self-explanatory. Will update the doc.

Reply:
Will there be other topology constraints? If no, how about just "TopologyConstraint".

If yes, how about "TopologyConstraints" with a nested "Spread" section?